FEATURE: Support for publishing topics as pages

eviltrout opened the pull request:
If the feature is enabled, staff members can construct a URL and publish a topic for others to browse without the regular Discourse chrome.

This is useful if you want to use Discourse like a CMS and publish topics as articles, which can then be embedded into other systems.

How to use:

1. activate `enable page publishing` in settings.

2. On a top, click this in the OP: 
![image](https://user-images.githubusercontent.com/17538/78600625-3d8b1380-7821-11ea-8264-6c00e8ce565e.png)

3. Build your custom URL:
![image](https://user-images.githubusercontent.com/17538/78600717-762aed00-7821-11ea-8a70-529d615a949b.png)

4. Your article will be published and you can visit it and read with no Discourse chrome:

![image](https://user-images.githubusercontent.com/17538/78600767-9195f800-7821-11ea-9510-17cbf35d6c1f.png)

jjaffeux commented on db/migrate/20200401172023_create_published_pages.rb:
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class CreatePublishedPages < ActiveRecord::Migration[6.0]
+  def change
```
given it doesn't soon like we use the id of this model maybe we could get rid of the auto increment and store the topic_id  in the id column as primary_key

eviltrout replied:
I thought of this but I considered in the future we might be publishing other things and I didn't want to "lock in" to a topic just yet.

For now the extra data for the key is minimal and I think is OK.

jjaffeux commented on spec/requests/published_pages_controller_spec.rb:
```diff
@@ -0,0 +1,176 @@
+      it "returns true for a new slug" do
+        get "/pub/check-slug.json?slug=cool-slug-man"
+        expect(response).to be_successful
+        json = JSON.parse(response.body)
```
I think you can use `response.parsed_body` instead of this

coding-horror commented:
Is there a visual indicator on the topic itself that it is "published" and
at some URL? I think we need that. Like a blue "staff notice" at the top of
the topic perhaps? (If we already have it, disregard)

On Mon, Apr 6, 2020 at 1:14 PM Robin Ward <notifications@github.com> wrote:

> If the feature is enabled, staff members can construct a URL and publish a
> topic for others to browse without the regular Discourse chrome.
>
> This is useful if you want to use Discourse like a CMS and publish topics
> as articles, which can then be embedded into other systems.
>
> How to use:
>
>    1.
>
>    activate enable page publishing in settings.
>    2.
>
>    On a top, click this in the OP:
>    [image: image]
>    <https://user-images.githubusercontent.com/17538/78600625-3d8b1380-7821-11ea-8264-6c00e8ce565e.png>
>    3.
>
>    Build your custom URL:
>    [image: image]
>    <https://user-images.githubusercontent.com/17538/78600717-762aed00-7821-11ea-8a70-529d615a949b.png>
>    4.
>
>    Your article will be published and you can visit it and read with no
>    Discourse chrome:
>
> [image: image]
> <https://user-images.githubusercontent.com/17538/78600767-9195f800-7821-11ea-9510-17cbf35d6c1f.png>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/discourse/discourse/pull/9364
> Commit Summary
>
>    - FEATURE: Support for publishing topics as pages
>
> File Changes
>
>    - *A* app/assets/javascripts/discourse/adapters/published-page.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-67eb4f5dd4d1bed4231d370658780c73>
>    (9)
>    - *M* app/assets/javascripts/discourse/components/text-field.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-22b9798877fafee94735fe8ba7a00734>
>    (26)
>    - *A* app/assets/javascripts/discourse/controllers/publish-page.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-d1c096b31d8384f80b0ca413da4b9e3d>
>    (115)
>    - *M* app/assets/javascripts/discourse/lib/transform-post.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-72654231c96ac05724bda90a3f88e893>
>    (4)
>    - *A* app/assets/javascripts/discourse/models/published-page.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-2a98bda9208f2abb2d8982e767868292>
>    (8)
>    - *M* app/assets/javascripts/discourse/routes/topic.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-ffb6ec1936d56a35426a999f99e1e05a>
>    (8)
>    - *A* app/assets/javascripts/discourse/templates/modal/publish-page.hbs
>    <https://github.com/discourse/discourse/pull/9364/files#diff-e6a51499845b1cead713d778c56f5909>
>    (60)
>    - *M* app/assets/javascripts/discourse/templates/topic.hbs
>    <https://github.com/discourse/discourse/pull/9364/files#diff-df5001640f8d307f2e4ebb00d41038b4>
>    (3)
>    - *M* app/assets/javascripts/discourse/widgets/post-admin-menu.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-d35bcedf2b4acba391b297f352145ae7>
>    (9)
>    - *M* app/assets/stylesheets/common/base/modal.scss
>    <https://github.com/discourse/discourse/pull/9364/files#diff-c515869c3e6a3d7c56858d9f0a30cca0>
>    (21)
>    - *A* app/assets/stylesheets/publish.scss
>    <https://github.com/discourse/discourse/pull/9364/files#diff-b8c49858cb4b4915f2755f0cf727877a>
>    (33)
>    - *A* app/controllers/published_pages_controller.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-2366a47bfd003299afab9dfffa5df997>
>    (58)
>    - *A* app/models/published_page.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-4b469baae007db35af1c607a29383117>
>    (46)
>    - *M* app/models/user_history.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-0a5ad2d838fbffeb0848dda1be5dbd12>
>    (8)
>    - *A* app/serializers/published_page_serializer.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-eda5784db26211fbe3f995a59d3e71bc>
>    (9)
>    - *M* app/serializers/topic_view_details_serializer.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-d969c655b27e6993b774723db8bea201>
>    (7)
>    - *M* app/serializers/topic_view_serializer.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-756f6795c82dac69125bd5b0f9c0366d>
>    (2)
>    - *M* app/services/staff_action_logger.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-fe92c3ce2e7290deb3cb1816fa3d4101>
>    (16)
>    - *A* app/views/layouts/publish.html.erb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-5a1147e0a0d96b11873637dd9eba7972>
>    (11)
>    - *A* app/views/published_pages/show.html.erb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-53b5af07b6dd2f48dc205fbc3e10c80f>
>    (18)
>    - *M* config/locales/client.en.yml
>    <https://github.com/discourse/discourse/pull/9364/files#diff-7ac2736614f0a5678e71cecf1563384e>
>    (15)
>    - *M* config/locales/server.en.yml
>    <https://github.com/discourse/discourse/pull/9364/files#diff-aeb8651199334c1cb831e884ff85773a>
>    (7)
>    - *M* config/routes.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-21497849d8f00507c9c8dcaf6288b136>
>    (6)
>    - *M* config/site_settings.yml
>    <https://github.com/discourse/discourse/pull/9364/files#diff-8f55b413f98b526e5dbb98a08945de89>
>    (2)
>    - *A* db/migrate/20200401172023_create_published_pages.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-cef758dd238f75bf0961e87ab405e7a6>
>    (13)
>    - *M* lib/guardian.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-e46fb5d4a3b5d6c6c5366683688a4836>
>    (8)
>    - *M* lib/stylesheet/watcher.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-464ec15909c602b73ee750900c9e04ea>
>    (2)
>    - *M* spec/components/guardian_spec.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-7d188edbba40e2d6a0b7569f319b703f>
>    (32)
>    - *A* spec/fabricators/published_page_fabricator.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-265bf6e32121d8763fc82a3ac06a555c>
>    (6)
>    - *A* spec/models/published_page_spec.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-74dfbda4e80dc676149b8b0d893f52ec>
>    (27)
>    - *A* spec/requests/published_pages_controller_spec.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-f407a17ab5850b0ea457124b98212a1d>
>    (176)
>    - *M* spec/serializers/topic_view_serializer_spec.rb
>    <https://github.com/discourse/discourse/pull/9364/files#diff-5790c3234f6398446eb0fc1ebf01245a>
>    (10)
>    - *A* test/javascripts/acceptance/page-publishing-test.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-c168c19280e577507dd8301151527913>
>    (40)
>    - *M* test/javascripts/fixtures/topic.js
>    <https://github.com/discourse/discourse/pull/9364/files#diff-7836945965fa69abb7a2496bfaddf223>
>    (1)
>
> Patch Links:
>
>    - https://github.com/discourse/discourse/pull/9364.patch
>    - https://github.com/discourse/discourse/pull/9364.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/discourse/discourse/pull/9364>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALTWVKIXN5LQX3SL4MR3MLRLIZZHANCNFSM4MCSEM2A>
> .
>

Abosnd02 commented:
aly0506

Abosnd02 commented:
> aly0506

eviltrout commented on app/assets/javascripts/discourse/templates/topic.hbs:
```diff
@@ -231,7 +231,8 @@
                 fillGapBefore=(action "fillGapBefore")
                 fillGapAfter=(action "fillGapAfter")
-                showInvite=(route-action "showInvite")}}
+                showInvite=(route-action "showInvite")
```
Noted thanks!

eviltrout commented on app/assets/javascripts/discourse/widgets/post-admin-menu.js:
```diff
@@ -120,6 +120,15 @@ export function buildManageButtons(attrs, currentUser, siteSettings) {
     }
   }
 
+  if (attrs.canPublishPage) {
```
Good catch! Amazed I missed this.

eviltrout commented on app/assets/javascripts/discourse/models/published-page.js:
```diff
@@ -0,0 +1,8 @@
+
+export default RestModel.extend({
+  url: computed("slug", function() {
+    return `${Discourse.BaseUrl}/pub/${this.slug}`;
```
This does not return the full URL with protocol and such.

jjaffeux commented:
So tried it this morning, and first as I said in chat this is very cool, I absolutely love the simplicity of this ❤️ I can't think of a simpler workflow.

I noticed a small issue, when adding a space at the beginning of my slug, it's not considered invalid:

<img width="683" alt="Screenshot 2020-04-08 at 09 17 10" src="https://user-images.githubusercontent.com/339945/78756506-32e49180-797b-11ea-86fb-271de0df3019.png">

And then when I will try to save it, it will say slug is taken.

I was also wondering, if we should handle category/tags ? at least optionally.

Another question I have is about duplicating content which is a very bad thing in SEO, we should have a story on this. As today we could have a public topic and a public CMS page with the same content.

eviltrout commented:
@coding-horror I added a section on top of the topic for staff members, that was a good call.

@jjaffeux I fixed that weird bug. I also added a rel="canonical" link to the page so that google will know the topic is the source. As for showing categories/tags I think we should investigate that in a follow up that perhaps allows users to customize the template.

k4rtik commented:
I just tried this, it seems like a cool and useful feature!

I have one question: do you think the publishing of non-public topics should be allowed? I tried that for a topic on my instance, and notice that I can view the page if I am logged in, but it is blank on a private window. I think if it is published, the rendering should be publicly accessible (even though the source stays private).

Personally, I am in favor of publishing a non-public topic as long as there is a visual indicator for the members who have access to that topic that it is also publicly published (and is not private to a category/group as they may have assumed).

eviltrout commented:
> I have one question: do you think the publishing of non-public topics should be allowed?

Yes the thinking is that people might use this as part of another system using iframes or API. In that case, only people who have access to the published page should be able to see it.

If we were to make private content public we'd have to be very sure the user wanted to do that.

hellekin commented:
I crave this feature! 

I guess the admin interface should add a "Published Pages" menu to change the CSS next to "Common", "Desktop", and "Mobile". Otherwise there's no obvious way to customize the `publish.scss`.

Use-case: the forum does not share the website's style. But the published pages should, as much as possible.

-----

I tried "publishing" a Staff topic but it failed, giving a white page. It would be nice to have, since this process works for banners (create the banner in #staff, and "banner topic").

-----

I tried publishing a page and using Firefox's "reader" feature: but the mini-profiler is taken into account, not the main text. Disable mini-profiler on published pages?
